marlowe/src/Language/Marlowe/SemanticsDeserialisation.hs

```diff
@@ -0,0 +1,480 @@
+module Language.Marlowe.SemanticsDeserialisation (byteStringToContract) where
+
+import           Language.Marlowe.Deserialisation (byteStringToInt, byteStringToList, byteStringToPositiveInt,
+                                                   getByteString)
+import           Language.Marlowe.SemanticsTypes  (Action (..), Bound (..), Case (..), ChoiceId (..), Contract (..),
+                                                   Observation (..), Party (..), Payee (..), Token (..),
+                                                   Value (AddValue, AvailableMoney, ChoiceValue, Cond, Constant, DivValue, MulValue, NegValue, Scale, SlotIntervalEnd, SlotIntervalStart, SubValue, UseValue),
+                                                   ValueId (..))
+import           Ledger                           (PubKeyHash (..), Slot (..))
+import           Ledger.Value                     (CurrencySymbol (..), TokenName (..))
+import           PlutusTx.Builtins                (BuiltinByteString)
+import           PlutusTx.Prelude                 ((%))
+
+
+byteStringToParty :: BuiltinByteString -> Maybe (Party, BuiltinByteString)
+byteStringToParty x =
+  case byteStringToPositiveInt x of
+    Nothing -> Nothing
+    Just (y, t1) ->
+      ( case getByteString t1 of
+          Nothing -> Nothing
+          Just (z, t2) ->
+            ( if y == 0
+                then Just (PK (PubKeyHash z), t2)
+                else
+                  ( if y == 1
+                      then Just (Role (TokenName z), t2)
+                      else Nothing
+                  )
+            )
+      )
+
+byteStringToChoiceId :: BuiltinByteString -> Maybe (ChoiceId, BuiltinByteString)
+byteStringToChoiceId x =
+  case getByteString x of
+    Nothing -> Nothing
+    Just (cn, t1) ->
+      ( case byteStringToParty t1 of
+          Nothing       -> Nothing
+          Just (co, t2) -> Just (ChoiceId cn co, t2)
+      )
+
+byteStringToValueId :: BuiltinByteString -> Maybe (ValueId, BuiltinByteString)
+byteStringToValueId x = case getByteString x of
+  Nothing     -> Nothing
+  Just (n, t) -> Just (ValueId n, t)
+
+byteStringToToken :: BuiltinByteString -> Maybe (Token, BuiltinByteString)
+byteStringToToken x =
+  case getByteString x of
+    Nothing -> Nothing
+    Just (cs, t1) ->
+      ( case getByteString t1 of
+          Nothing       -> Nothing
+          Just (tn, t2) -> Just (Token (CurrencySymbol cs) (TokenName tn), t2)
+      )
+
+byteStringToObservation :: BuiltinByteString -> Maybe (Observation, BuiltinByteString)
+byteStringToObservation x =
+  case byteStringToPositiveInt x of
+    Nothing -> Nothing
+    Just (y, t1) ->
```

Comment:
Doing straightforward `case of` should result in almost identically optimized code generated by Haskell pattern-matching optimizer. It also would be simpler to add/remove constructors.